{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARCHS4_Cell-lines',\n",
       " 'ARCHS4_IDG_Coexp',\n",
       " 'ARCHS4_Kinases_Coexp',\n",
       " 'ARCHS4_TFs_Coexp',\n",
       " 'ARCHS4_Tissues',\n",
       " 'Achilles_fitness_decrease',\n",
       " 'Achilles_fitness_increase',\n",
       " 'Aging_Perturbations_from_GEO_down',\n",
       " 'Aging_Perturbations_from_GEO_up',\n",
       " 'Allen_Brain_Atlas_down',\n",
       " 'Allen_Brain_Atlas_up',\n",
       " 'BioCarta_2013',\n",
       " 'BioCarta_2015',\n",
       " 'BioCarta_2016',\n",
       " 'BioPlanet_2019',\n",
       " 'BioPlex_2017',\n",
       " 'CORUM',\n",
       " 'Cancer_Cell_Line_Encyclopedia',\n",
       " 'ChEA_2013',\n",
       " 'ChEA_2015',\n",
       " 'ChEA_2016',\n",
       " 'Chromosome_Location',\n",
       " 'Chromosome_Location_hg19',\n",
       " 'ClinVar_2019',\n",
       " 'DSigDB',\n",
       " 'Data_Acquisition_Method_Most_Popular_Genes',\n",
       " 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
       " 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
       " 'DisGeNET',\n",
       " 'Disease_Perturbations_from_GEO_down',\n",
       " 'Disease_Perturbations_from_GEO_up',\n",
       " 'Disease_Signatures_from_GEO_down_2014',\n",
       " 'Disease_Signatures_from_GEO_up_2014',\n",
       " 'DrugMatrix',\n",
       " 'Drug_Perturbations_from_GEO_2014',\n",
       " 'Drug_Perturbations_from_GEO_down',\n",
       " 'Drug_Perturbations_from_GEO_up',\n",
       " 'ENCODE_Histone_Modifications_2013',\n",
       " 'ENCODE_Histone_Modifications_2015',\n",
       " 'ENCODE_TF_ChIP-seq_2014',\n",
       " 'ENCODE_TF_ChIP-seq_2015',\n",
       " 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
       " 'ESCAPE',\n",
       " 'Enrichr_Libraries_Most_Popular_Genes',\n",
       " 'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
       " 'Epigenomics_Roadmap_HM_ChIP-seq',\n",
       " 'GO_Biological_Process_2013',\n",
       " 'GO_Biological_Process_2015',\n",
       " 'GO_Biological_Process_2017',\n",
       " 'GO_Biological_Process_2017b',\n",
       " 'GO_Biological_Process_2018',\n",
       " 'GO_Cellular_Component_2013',\n",
       " 'GO_Cellular_Component_2015',\n",
       " 'GO_Cellular_Component_2017',\n",
       " 'GO_Cellular_Component_2017b',\n",
       " 'GO_Cellular_Component_2018',\n",
       " 'GO_Molecular_Function_2013',\n",
       " 'GO_Molecular_Function_2015',\n",
       " 'GO_Molecular_Function_2017',\n",
       " 'GO_Molecular_Function_2017b',\n",
       " 'GO_Molecular_Function_2018',\n",
       " 'GTEx_Tissue_Sample_Gene_Expression_Profiles_down',\n",
       " 'GTEx_Tissue_Sample_Gene_Expression_Profiles_up',\n",
       " 'GWAS_Catalog_2019',\n",
       " 'GeneSigDB',\n",
       " 'Gene_Perturbations_from_GEO_down',\n",
       " 'Gene_Perturbations_from_GEO_up',\n",
       " 'Genes_Associated_with_NIH_Grants',\n",
       " 'Genome_Browser_PWMs',\n",
       " 'HMDB_Metabolites',\n",
       " 'HomoloGene',\n",
       " 'HumanCyc_2015',\n",
       " 'HumanCyc_2016',\n",
       " 'Human_Gene_Atlas',\n",
       " 'Human_Phenotype_Ontology',\n",
       " 'InterPro_Domains_2019',\n",
       " 'Jensen_COMPARTMENTS',\n",
       " 'Jensen_DISEASES',\n",
       " 'Jensen_TISSUES',\n",
       " 'KEA_2013',\n",
       " 'KEA_2015',\n",
       " 'KEGG_2013',\n",
       " 'KEGG_2015',\n",
       " 'KEGG_2016',\n",
       " 'KEGG_2019_Human',\n",
       " 'KEGG_2019_Mouse',\n",
       " 'Kinase_Perturbations_from_GEO_down',\n",
       " 'Kinase_Perturbations_from_GEO_up',\n",
       " 'LINCS_L1000_Chem_Pert_down',\n",
       " 'LINCS_L1000_Chem_Pert_up',\n",
       " 'LINCS_L1000_Kinase_Perturbations_down',\n",
       " 'LINCS_L1000_Kinase_Perturbations_up',\n",
       " 'LINCS_L1000_Ligand_Perturbations_down',\n",
       " 'LINCS_L1000_Ligand_Perturbations_up',\n",
       " 'Ligand_Perturbations_from_GEO_down',\n",
       " 'Ligand_Perturbations_from_GEO_up',\n",
       " 'MCF7_Perturbations_from_GEO_down',\n",
       " 'MCF7_Perturbations_from_GEO_up',\n",
       " 'MGI_Mammalian_Phenotype_2013',\n",
       " 'MGI_Mammalian_Phenotype_2017',\n",
       " 'MGI_Mammalian_Phenotype_Level_3',\n",
       " 'MGI_Mammalian_Phenotype_Level_4',\n",
       " 'MGI_Mammalian_Phenotype_Level_4_2019',\n",
       " 'MSigDB_Computational',\n",
       " 'MSigDB_Oncogenic_Signatures',\n",
       " 'Microbe_Perturbations_from_GEO_down',\n",
       " 'Microbe_Perturbations_from_GEO_up',\n",
       " 'Mouse_Gene_Atlas',\n",
       " 'NCI-60_Cancer_Cell_Lines',\n",
       " 'NCI-Nature_2015',\n",
       " 'NCI-Nature_2016',\n",
       " 'NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions',\n",
       " 'NIH_Funded_PIs_2017_Human_AutoRIF',\n",
       " 'NIH_Funded_PIs_2017_Human_GeneRIF',\n",
       " 'NURSA_Human_Endogenous_Complexome',\n",
       " 'OMIM_Disease',\n",
       " 'OMIM_Expanded',\n",
       " 'Old_CMAP_down',\n",
       " 'Old_CMAP_up',\n",
       " 'PPI_Hub_Proteins',\n",
       " 'Panther_2015',\n",
       " 'Panther_2016',\n",
       " 'Pfam_Domains_2019',\n",
       " 'Pfam_InterPro_Domains',\n",
       " 'PheWeb_2019',\n",
       " 'Phosphatase_Substrates_from_DEPOD',\n",
       " 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO',\n",
       " 'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_AutoRIF_Gene_Lists',\n",
       " 'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
       " 'Rare_Diseases_GeneRIF_Gene_Lists',\n",
       " 'Reactome_2013',\n",
       " 'Reactome_2015',\n",
       " 'Reactome_2016',\n",
       " 'SILAC_Phosphoproteomics',\n",
       " 'SubCell_BarCode',\n",
       " 'SysMyo_Muscle_Gene_Sets',\n",
       " 'TF-LOF_Expression_from_GEO',\n",
       " 'TF_Perturbations_Followed_by_Expression',\n",
       " 'TRANSFAC_and_JASPAR_PWMs',\n",
       " 'TRRUST_Transcription_Factors_2019',\n",
       " 'TargetScan_microRNA',\n",
       " 'TargetScan_microRNA_2017',\n",
       " 'Tissue_Protein_Expression_from_Human_Proteome_Map',\n",
       " 'Tissue_Protein_Expression_from_ProteomicsDB',\n",
       " 'Transcription_Factor_PPIs',\n",
       " 'UK_Biobank_GWAS_v1',\n",
       " 'VirusMINT',\n",
       " 'Virus_Perturbations_from_GEO_down',\n",
       " 'Virus_Perturbations_from_GEO_up',\n",
       " 'WikiPathways_2013',\n",
       " 'WikiPathways_2015',\n",
       " 'WikiPathways_2016',\n",
       " 'WikiPathways_2019_Human',\n",
       " 'WikiPathways_2019_Mouse',\n",
       " 'dbGaP',\n",
       " 'huMAP',\n",
       " 'miRTarBase_2017']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gp.get_library_name()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['TIMP1', 'CASP3', 'PLA2G5', 'CDC6', 'TIMP4', 'INHBB', 'HOMER1',\n",
    "       'PDLIM3', 'RPRM', 'DKK1', 'DYNLT3', 'NET1', 'NET1', 'CCNA1', 'PDGFRA',\n",
    "       'NSUN5', 'NSUN5', 'PYGB', 'SLC43A3', 'TAGLN2', 'CDK2', 'ID4', 'ATRX',\n",
    "       'SH2B2', 'SFRP4', 'FAM3C', 'ACOX2', 'THBS2', 'ZNF609', 'SFRP1',\n",
    "       'IGFBP3', 'MGMT', 'FZD3', 'CCNE1', 'LGALS8', 'LGALS8', 'TP53', 'THBS3',\n",
    "       'IDH1', 'MDK', 'DCTD', 'PRKAR2B', 'KLF13', 'POLE3', 'IGFBP2', 'IGFBP2',\n",
    "       'CCND1', 'CDKN2A', 'PDPN', 'ADM', 'FZD7', 'GHR', 'EZH2', 'TFRC',\n",
    "       'RGS14', 'PIK3CA', 'VEGFA', 'REPS2', 'MT1M', 'CDKN1C', 'STMN1', 'IRS2',\n",
    "       'HSPA1B', 'LIFR', 'G6PC3', 'GADD45G', 'FBXO17', 'PTPRN', 'WAC',\n",
    "       'STAG3L4', 'EFEMP2', 'EFEMP2', 'PPARGC1A', 'EGFR', 'MT1E']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "enr = gp.enrichr(gene_list=genelist,\n",
    "                 gene_sets=['GO_Biological_Process_2018',\n",
    "                            'GO_Cellular_Component_2018',\n",
    "                            'GO_Molecular_Function_2018'],\n",
    "#                  outdir='test/enrichr_kegg',\n",
    "                 no_plot=True,\n",
    "                 cutoff=0.001 # test dataset, use lower value from range(0,1)\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       0.035781\n",
       "1       0.044900\n",
       "2      -0.012389\n",
       "3       0.037923\n",
       "4      -0.010752\n",
       "          ...   \n",
       "3080   -0.013115\n",
       "3081   -0.007255\n",
       "3082    0.000888\n",
       "3083   -0.013590\n",
       "3084   -0.000256\n",
       "Name: PHATE 1, Length: 3085, dtype: float64"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tumor_dfs[\"PJ016\"].obs[\"PHATE 1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000021</td>\n",
       "      <td>119.406906</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>PDGFRA;CDKN2A;IGFBP3;LIFR;IRS2;CDC6;EGFR;VEGFA...</td>\n",
       "      <td>6.177606</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16/740</td>\n",
       "      <td>4.032042e-09</td>\n",
       "      <td>regulation of cell proliferation (GO:0042127)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000029</td>\n",
       "      <td>139.994319</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>PDGFRA;FZD3;MGMT;DKK1;EGFR;VEGFA;SFRP1;PIK3CA;...</td>\n",
       "      <td>7.658321</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13/485</td>\n",
       "      <td>1.151025e-08</td>\n",
       "      <td>negative regulation of apoptotic process (GO:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000044</td>\n",
       "      <td>125.098116</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>PDGFRA;LIFR;IRS2;CDC6;DKK1;EGFR;VEGFA;SFRP1;AC...</td>\n",
       "      <td>7.156620</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13/519</td>\n",
       "      <td>2.561581e-08</td>\n",
       "      <td>positive regulation of cellular process (GO:00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000036</td>\n",
       "      <td>1985.799593</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>CCND1;CDK2;WAC;TP53</td>\n",
       "      <td>114.285714</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4/10</td>\n",
       "      <td>2.843212e-08</td>\n",
       "      <td>G1 DNA damage checkpoint (GO:0044783)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000084</td>\n",
       "      <td>310.781696</td>\n",
       "      <td>GO_Biological_Process_2018</td>\n",
       "      <td>CCNA1;CCND1;CDKN2A;CCNE1;POLE3;CDK2;CDC6</td>\n",
       "      <td>19.047619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7/105</td>\n",
       "      <td>8.204163e-08</td>\n",
       "      <td>G1/S transition of mitotic cell cycle (GO:0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.459770</td>\n",
       "      <td>GO_Molecular_Function_2018</td>\n",
       "      <td>TP53</td>\n",
       "      <td>1.006036</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/284</td>\n",
       "      <td>6.331734e-01</td>\n",
       "      <td>transcriptional activator activity, RNA polyme...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.431396</td>\n",
       "      <td>GO_Molecular_Function_2018</td>\n",
       "      <td>TP53</td>\n",
       "      <td>0.978474</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/292</td>\n",
       "      <td>6.434657e-01</td>\n",
       "      <td>transcription regulatory region sequence-speci...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.206478</td>\n",
       "      <td>GO_Molecular_Function_2018</td>\n",
       "      <td>PPARGC1A</td>\n",
       "      <td>0.725163</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/394</td>\n",
       "      <td>7.522146e-01</td>\n",
       "      <td>sequence-specific DNA binding (GO:0043565)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1378</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.177485</td>\n",
       "      <td>GO_Molecular_Function_2018</td>\n",
       "      <td>FBXO17</td>\n",
       "      <td>0.685166</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/417</td>\n",
       "      <td>7.717931e-01</td>\n",
       "      <td>ubiquitin-protein transferase activity (GO:000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.135197</td>\n",
       "      <td>GO_Molecular_Function_2018</td>\n",
       "      <td>TP53</td>\n",
       "      <td>0.621118</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1/460</td>\n",
       "      <td>8.043929e-01</td>\n",
       "      <td>RNA polymerase II regulatory region sequence-s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1380 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Adjusted P-value  Combined Score                    Gene_set  \\\n",
       "0             0.000021      119.406906  GO_Biological_Process_2018   \n",
       "1             0.000029      139.994319  GO_Biological_Process_2018   \n",
       "2             0.000044      125.098116  GO_Biological_Process_2018   \n",
       "3             0.000036     1985.799593  GO_Biological_Process_2018   \n",
       "4             0.000084      310.781696  GO_Biological_Process_2018   \n",
       "...                ...             ...                         ...   \n",
       "1375          1.000000        0.459770  GO_Molecular_Function_2018   \n",
       "1376          1.000000        0.431396  GO_Molecular_Function_2018   \n",
       "1377          1.000000        0.206478  GO_Molecular_Function_2018   \n",
       "1378          1.000000        0.177485  GO_Molecular_Function_2018   \n",
       "1379          1.000000        0.135197  GO_Molecular_Function_2018   \n",
       "\n",
       "                                                  Genes  Odds Ratio  \\\n",
       "0     PDGFRA;CDKN2A;IGFBP3;LIFR;IRS2;CDC6;EGFR;VEGFA...    6.177606   \n",
       "1     PDGFRA;FZD3;MGMT;DKK1;EGFR;VEGFA;SFRP1;PIK3CA;...    7.658321   \n",
       "2     PDGFRA;LIFR;IRS2;CDC6;DKK1;EGFR;VEGFA;SFRP1;AC...    7.156620   \n",
       "3                                   CCND1;CDK2;WAC;TP53  114.285714   \n",
       "4              CCNA1;CCND1;CDKN2A;CCNE1;POLE3;CDK2;CDC6   19.047619   \n",
       "...                                                 ...         ...   \n",
       "1375                                               TP53    1.006036   \n",
       "1376                                               TP53    0.978474   \n",
       "1377                                           PPARGC1A    0.725163   \n",
       "1378                                             FBXO17    0.685166   \n",
       "1379                                               TP53    0.621118   \n",
       "\n",
       "      Old Adjusted P-value  Old P-value Overlap       P-value  \\\n",
       "0                        0            0  16/740  4.032042e-09   \n",
       "1                        0            0  13/485  1.151025e-08   \n",
       "2                        0            0  13/519  2.561581e-08   \n",
       "3                        0            0    4/10  2.843212e-08   \n",
       "4                        0            0   7/105  8.204163e-08   \n",
       "...                    ...          ...     ...           ...   \n",
       "1375                     0            0   1/284  6.331734e-01   \n",
       "1376                     0            0   1/292  6.434657e-01   \n",
       "1377                     0            0   1/394  7.522146e-01   \n",
       "1378                     0            0   1/417  7.717931e-01   \n",
       "1379                     0            0   1/460  8.043929e-01   \n",
       "\n",
       "                                                   Term  \n",
       "0         regulation of cell proliferation (GO:0042127)  \n",
       "1     negative regulation of apoptotic process (GO:0...  \n",
       "2     positive regulation of cellular process (GO:00...  \n",
       "3                 G1 DNA damage checkpoint (GO:0044783)  \n",
       "4     G1/S transition of mitotic cell cycle (GO:0000...  \n",
       "...                                                 ...  \n",
       "1375  transcriptional activator activity, RNA polyme...  \n",
       "1376  transcription regulatory region sequence-speci...  \n",
       "1377         sequence-specific DNA binding (GO:0043565)  \n",
       "1378  ubiquitin-protein transferase activity (GO:000...  \n",
       "1379  RNA polymerase II regulatory region sequence-s...  \n",
       "\n",
       "[1380 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enr.results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
